src/window.cpp

```diff
@@ -895,57 +874,100 @@ static bool MayBeShown(const Window *w)
  * @param top Top edge of the rectangle that should be repainted
  * @param right Right edge of the rectangle that should be repainted
  * @param bottom Bottom edge of the rectangle that should be repainted
+ * @param gfx_dirty Whether to mark gfx dirty
  */
-static void DrawOverlappedWindow(Window *w, int left, int top, int right, int bottom)
-{
-	const Window *v;
-	FOR_ALL_WINDOWS_FROM_BACK_FROM(v, w->z_front) {
-		if (MayBeShown(v) &&
-				right > v->left &&
-				bottom > v->top &&
-				left < v->left + v->width &&
-				top < v->top + v->height) {
-			/* v and rectangle intersect with each other */
-			int x;
-
-			if (left < (x = v->left)) {
-				DrawOverlappedWindow(w, left, top, x, bottom);
-				DrawOverlappedWindow(w, x, top, right, bottom);
-				return;
+void DrawOverlappedWindow(Window *w, const int left, const int top, const int right, const int bottom, const bool gfx_dirty)
+{
+	std::function<void(Window const * const, int, int, int, int)> draw_overlapped;
+	draw_overlapped = [&w, &gfx_dirty, &draw_overlapped](Window const * const from_w, int left, int top, int right, int bottom){
+		const Window *v;
+		FOR_ALL_WINDOWS_FROM_BACK_FROM(v, from_w) {
+			if (!MayBeShown(v)) {
+				/* Window 'v' is not visible, no occulsion possible. */
+				continue;
 			}
 
-			if (right > (x = v->left + v->width)) {
-				DrawOverlappedWindow(w, left, top, x, bottom);
-				DrawOverlappedWindow(w, x, top, right, bottom);
-				return;
```

Comment:
Although the compiler _might_ be smart enough to eliminate the tail-recursive call, why not just fall through to the next test and never have to worry about `v` again?